### PR TITLE
[Java] Fix multiple `FunctionManager`s creating multiple `ClassLoader `s

### PR DESCRIPTION
**Pull request by kfstorm**
## Why are these changes needed?

This bug only occurs if `ray.job.resource-path` is set.

In multi-threading scenario, different `RayNativeRuntime` instances holds different `FunctionManager` instances, which means different `ClassLoader` instances for the same job. If the user code involves any static variables, it may behave strangely.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: java/runtime/src/main/java/org/ray/runtime/functionmanager/FunctionManager.java [hunk 1]
```diff
@@ -92,31 +100,37 @@ public RayFunction getFunction(JobId jobId, RayFunc func) {
...
-        File resourceDir = new File(jobResourcePath + "/" + jobId.toString() + "/");
-        Collection<File> files = FileUtils.listFiles(resourceDir,
-            new RegexFileFilter(".*\\.jar"), DirectoryFileFilter.DIRECTORY);
-        files.add(resourceDir);
-        final List<URL> urlList = files.stream().map(file -> {
-          try {
-            return file.toURI().toURL();
-          } catch (MalformedURLException e) {
-            throw new RuntimeException(e);
+      synchronized (jobFunctionTablesLock) {
```
Comment (raulchen) [inline java/runtime/src/main/java/org/ray/runtime/functionmanager/FunctionManager.java]: just `synchronized (this)`?

File: java/runtime/src/main/java/org/ray/runtime/functionmanager/FunctionManager.java [hunk 2]
```diff
@@ -92,31 +100,37 @@ public RayFunction getFunction(JobId jobId, RayFunc func) {
...
+        jobFunctionTable = jobFunctionTables.get(jobId);
+        if (jobFunctionTable == null) {
```
Comment (raulchen) [inline java/runtime/src/main/java/org/ray/runtime/functionmanager/FunctionManager.java]: maybe wrap following code in a method (e.g., `createJobFunctionTable`) so the structure is clearer?

File: java/runtime/src/main/java/org/ray/runtime/functionmanager/FunctionManager.java (hunk 1 above)
Comment (kfstorm) [reply java/runtime/src/main/java/org/ray/runtime/functionmanager/FunctionManager.java]: Done.

File: java/runtime/src/main/java/org/ray/runtime/functionmanager/FunctionManager.java (hunk 2 above)
Comment (kfstorm) [reply java/runtime/src/main/java/org/ray/runtime/functionmanager/FunctionManager.java]: Done.

**Review by raulchen — APPROVED** [review]
LGTM. Please fix another `synchronized(this)` issue.

**Review by zhijunfu — APPROVED** [review]

File: java/runtime/src/main/java/org/ray/runtime/functionmanager/FunctionManager.java [hunk 3]
```diff
@@ -129,22 +145,29 @@ public RayFunction getFunction(JobId jobId, JavaFunctionDescriptor functionDescr
...
-      this.functions = new HashMap<>();
+      this.functions = new ConcurrentHashMap<>();
     }
 
     RayFunction getFunction(JavaFunctionDescriptor descriptor) {
       Map<Pair<String, String>, RayFunction> classFunctions = functions.get(descriptor.className);
       if (classFunctions == null) {
-        classFunctions = loadFunctionsForClass(descriptor.className);
-        functions.put(descriptor.className, classFunctions);
+        synchronized (lock) {
```
[the remark this answers is not in the document]
Comment (kfstorm) [reply java/runtime/src/main/java/org/ray/runtime/functionmanager/FunctionManager.java]: Ah, right! Code updated.

**Pull request merged**